daphne/src/testing.rs

```diff
@@ -272,3 +303,150 @@ impl DapLeader<BearerToken> for MockAggregator {
+macro_rules! checked_process {
+    ($store:expr, $nonce:expr) => {{
+        let observed = $store
+            .processed
+            .lock()
+            .map_err(int_err)?
+            .deref()
+            .contains($nonce);
+        if observed && !$store.collected.lock().map_err(int_err)?.deref() {
+            return Err(TransitionFailure::ReportReplayed);
+        } else if !observed && *$store.collected.lock().map_err(int_err)?.deref() {
+            return Err(TransitionFailure::BatchCollected);
+        }
+        $store
+            .processed
+            .lock()
+            .map_err(int_err)?
+            .deref_mut()
+            .insert($nonce.clone());
+    }};
+}
+
+// TODO(nakatsuka-y remove dead_code once all functions are used)
+#[allow(dead_code)]
+impl ReportStore {
+    pub fn new() -> Self {
+        let pending: Vec<Report> = Vec::new();
+        let processed: HashSet<Nonce> = HashSet::new();
+        let collected = false;
+
+        Self {
+            pending: Arc::new(Mutex::new(pending)),
+            processed: Arc::new(Mutex::new(processed)),
+            collected: Arc::new(Mutex::new(collected)),
+        }
+    }
+
+    pub async fn process_delete_all(&self) -> Result<(), TransitionFailure> {
+        self.pending.lock().map_err(int_err)?.deref_mut().clear();
+        self.processed.lock().map_err(int_err)?.deref_mut().clear();
+        *self.collected.lock().map_err(int_err)?.deref_mut() = false;
+        Ok(())
+    }
+
+    pub async fn process_get_pending(
+        &self,
+        info: ReportStoreGetPending,
+    ) -> Result<Vec<Report>, TransitionFailure> {
+        let reports_drained = std::cmp::min(
+            info.reports_requested.try_into().unwrap(),
+            self.pending.lock().map_err(int_err)?.deref().len(),
+        );
+        let reports: Vec<Report> = self
+            .pending
+            .lock()
+            .map_err(int_err)?
+            .deref_mut()
+            .drain(..reports_drained)
+            .collect();
+        Ok(reports)
+    }
+
+    pub async fn process_put_pending(&self, report: Report) -> Result<(), TransitionFailure> {
+        checked_process!(self, &report.nonce);
+        self.pending
+            .lock()
+            .map_err(int_err)?
+            .deref_mut()
+            .push(report);
+        Ok(())
+    }
+
+    pub async fn process_put_processed(&self, nonce: Nonce) -> Result<(), TransitionFailure> {
+        checked_process!(self, &nonce);
+        Ok(())
+    }
+
+    pub async fn process_mark_collected(&self) -> Result<(), TransitionFailure> {
+        *self.collected.lock().map_err(int_err)?.deref_mut() = true;
+        Ok(())
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+pub(crate) struct HelperStateStoreGetPending {
+    pub(crate) reports_requested: u64,
+}
+
+// TODO(nakatsuka-y remove dead_code once all functions are used)
+#[allow(dead_code)]
+pub(crate) struct HelperStateStore {
+    helper_state: Arc<Mutex<Option<DapHelperState>>>,
+}
```

Comment:
Two comments here:
* Only one helper state blob is stored at any given time. However, it's possible that the helper will be handling multiple aggregation flows simultaneously. The data type here should be something like a `HashMap<Id, DapHelperState>` that maps [aggregation job IDs](https://ietf-wg-ppm.github.io/draft-ietf-ppm-dap/draft-ietf-ppm-dap.html#section-4.3.1) to state blobs. 
* The contents of this struct are fairly trivial. How about doing away with `HelperStateStore` and moving the `Arc<Mutex<_>>` into `MockAggregator`?

Reply:
Good point, I've moved `HelperStateStore` directly into `MockAggregator` and mapped aggregation job IDs to `DapHelperState`.